client/my-sites/site-settings/form-general.jsx

```diff
@@ -113,57 +119,136 @@ const FormGeneral = React.createClass( {
 			amp_is_supported: false,
 			amp_is_enabled: false,
 		} );
-	},
+		this.props.clearDirtyFields();
+		this.props.updateFields( this.getFormSettings( this.props.settings ) );
+	}
 
-	onTimezoneSelect( timezone ) {
-		this.setState( { timezone_string: timezone } );
-	},
+	componentWillReceiveProps( nextProps ) {
+		this._showWarning( nextProps.site );
+
+		if ( nextProps.settings !== this.props.settings ) {
+			let newState = this.getFormSettings( nextProps.settings );
+			//If we have any fields that the user has updated,
+			//do not wipe out those fields from the poll update.
+			newState = omit( newState, nextProps.dirtyFields );
+			nextProps.updateFields( newState );
+		}
+
+		if (
+			this.props.isSavingSettings &&
+			! nextProps.isSavingSettings &&
+			nextProps.isSaveRequestSuccessful
+		) {
+			nextProps.clearDirtyFields();
+			nextProps.markSaved();
+		}
+	}
 
 	onRecordEvent( eventAction ) {
 		return this.recordEvent.bind( this, eventAction );
-	},
+	}
 
 	onRecordEventOnce( key, eventAction ) {
 		return this.recordEventOnce.bind( this, key, eventAction );
-	},
+	}
+
+	recordEvent( eventAction ) {
+		debug( 'record event: %o', eventAction );
+		analytics.ga.recordEvent( 'Site Settings', eventAction );
+	}
+
+	/**
+	 * Record an analytics event only once per mounted component instance
+	 * @param  {string} key         - unique key to namespace the event
+	 * @param  {string} eventAction - the description of the action to appear in analytics
+	 */
+	recordEventOnce( key, eventAction ) {
+		debug( 'record event once: %o - %o', key, eventAction );
+		if ( this.state[ 'recordEventOnce-' + key ] ) {
+			return;
+		}
+		this.recordEvent( eventAction );
+		this.setState( { [ 'recordEventOnce-' + key ]: true } );
+	}
+
+	handleRadio = event => {
+		const currentTargetName = event.currentTarget.name,
+			currentTargetValue = event.currentTarget.value;
+
+		this.props.updateFields( { [ currentTargetName ]: currentTargetValue } );
+	};
+
+	handleSubmitForm = event => {
+		if ( ! event.isDefaultPrevented() && event.nativeEvent ) {
+			event.preventDefault();
+		}
+
+		this.submitForm();
+		this.recordEvent( 'Clicked Save Settings Button' );
+	};
+
+	submitForm() {
```

Comment:
I generally agree that it'd be nice if the form components didn't rely so much on helpers from `formBase`. I think between selectors and action creators we could probably limit the amount of duplicated code. Even with event recording we could take advantage of [existing middlewares and action extenders](https://github.com/Automattic/wp-calypso/tree/master/client/state/analytics), or create specialized action creators using [`extendAction`](https://github.com/Automattic/wp-calypso/blob/fcf789ebb43b7d8fe9a912d8fa82d5bee66573b5/client/state/README.md#extendaction-action-data-). Currently though this component reimplements `componentWillReceiveProps`, `recordEvent`, `recordEventOnce`, `handleRadio`, `handleSubmitForm`, and `submitForm`. It looks like you're already on the path to help reduce some of the event handler logic in #9867, and I think `handleSubmitForm` and `submitForm` could probably be collapsed into a single function (by the way, does this component need the `isDefaultPrevented` checks?).